[Gradient Compression] Check start_PowerSGD_iter > 1

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51427 [Gradient Compression] Check start_PowerSGD_iter > 1**

A user reported that `start_PowerSGD_iter` failed when it's set as 1. This is because allocating memory for error tensors somehow overlap with bucket rebuilding process at iteration 1.

Check `start_PowerSGD_iter > 1` instead of `start_PowerSGD_iter >= 1`.

Also added a unit test of `test_invalid_powerSGD_state` and some guidance on tuning PowerSGD configs.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26166897](https://our.internmc.facebook.com/intern/diff/D26166897/)

Commit:
[Gradient Compression] Check start_PowerSGD_iter > 1

A user reported that `start_PowerSGD_iter` failed when it's set as 1. This is because allocating memory for error tensors somehow overlap with bucket rebuilding process at iteration 1.

Assert `start_PowerSGD_iter > 1` instead of `start_PowerSGD_iter >= 1`.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26166897](https://our.internmc.facebook.com/intern/diff/D26166897/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Check start_PowerSGD_iter > 1"


A user reported that `start_PowerSGD_iter` failed when it's set as 1. This is because allocating memory for error tensors somehow overlap with bucket rebuilding process at iteration 1.

Assert `start_PowerSGD_iter > 1` instead of `start_PowerSGD_iter >= 1`.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26166897](https://our.internmc.facebook.com/intern/diff/D26166897/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
This diff doesn't seem to modify any of the asserts as the description suggests?

Commit:
Update on "[Gradient Compression] Check start_PowerSGD_iter > 1"


A user reported that `start_PowerSGD_iter` failed when it's set as 1. This is because allocating memory for error tensors somehow overlap with bucket rebuilding process at iteration 1.

Assert `start_PowerSGD_iter > 1` instead of `start_PowerSGD_iter >= 1`.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26166897](https://our.internmc.facebook.com/intern/diff/D26166897/)

[ghstack-poisoned]

SciPioneer commented:
> This diff doesn't seem to modify any of the asserts as the description suggests?

Updated. The diff was somehow messed up in the stack.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -80,7 +80,11 @@ def __init__(
         # which will complicate the implementations of error feedback and warm-up.
         # Running vanilla allreduce in the first few iterations can avoid this complexity.
-        assert start_powerSGD_iter >= 1
+        if use_error_feedback or warm_start:
```
Have we checked all internal callsites to ensure that this won't result in a crash for users who are using the hook? 
Also, would raising an exception here rather than an assert be more appropriate? Either way, would also be useful to add a test that ensures an error is appropriately raised when this path is hit.

SciPioneer replied:
> Have we checked all internal callsites to ensure that this won't result in a crash for users who are using the hook?

There is no internal callsite so far.

> Also, would raising an exception here rather than an assert be more appropriate? Either way, would also be useful to add a test that ensures an error is appropriately raised when this path is hit.

Replaced the assertion by exception, and added a unit test `test_invalid_powerSGD_state`.

rohan-varma reviewed (approved):
LGTM, though please consider adding the test and ensuring all CI checks pass

rohan-varma reviewed (commented):
Can we also add appropriate docstrings for these state classes for the hooks, and also ensure that subtleties such as the start iteration are well documented both in this state and in the actual hook? 

Added much more documentations, including the invalid value range and the guidance on tuning the PowerSGD configs.

Commit:
Update on "[Gradient Compression] Check start_PowerSGD_iter > 1"


A user reported that `start_PowerSGD_iter` failed when it's set as 1. This is because allocating memory for error tensors somehow overlap with bucket rebuilding process at iteration 1.

Check `start_PowerSGD_iter > 1` instead of `start_PowerSGD_iter >= 1`.

Also added a unit test of `test_invalid_powerSGD_state` and some guidance on tuning PowerSGD configs.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26166897](https://our.internmc.facebook.com/intern/diff/D26166897/)

[ghstack-poisoned]
